### ISSUE TITLE
dns hijacking bug in android 

### ISSUE THREAD

**Issue by mokhtarabadi** [eycorsican/leaf#180]
when I enabled domainResolve option in routing, I can't access block sites because my isp change the responded DNS
because DNS client create UDP socket and UDP socket protect from android VPN service

**Comment by mokhtarabadi** [comment]
@eycorsican I think this an important bug...

**Comment by eycorsican** [comment]
Don't expect a response if you're not trying your best to describe the issue.

**Comment by mokhtarabadi** [comment]
every UDP packet that leaf created will bypass the android VPN network [here](https://github.com/eycorsican/leaf/blob/23dc682e34803f0c148a3f47e46c4bdb709827c4/leaf/src/proxy/mod.rs#L314) 
when enable domainResolve property in router mean:
```json
"router" : {
    "domainResolve" : true
}
```
leaf trying to resolve hostnames through [dns_client](https://github.com/eycorsican/leaf/blob/master/leaf/src/app/dns_client.rs) which use this [line](https://github.com/eycorsican/leaf/blob/23dc682e34803f0c148a3f47e46c4bdb709827c4/leaf/src/proxy/mod.rs#L300) so DNS traffics will bypass the android VPN network, and ISPs can change the DNS server responses easily!

English isn't my native language...
hope you understand the problem

**Comment by eycorsican** [comment]
If you don't want those dns requests to be sent out directly, just try to put a dispatcher into the dns client and retrieve a socket with that for sending queries.

**Comment by mokhtarabadi** [comment]
Hey, @eycorsican I check the problem again and find these:
when resolveDomain property in the router is set to true, leaf try to resolve the IP of the domain directly by send DNS request to the DNS server (because in android every UDP packet that was created is protected from the VPN network or simply from tuntap) and ISP change the DNS server response and change the IP to 10.10.34.34 (Iran Firewall) so it's the wrong IP and if try to send DNS request through the leaf (mean comment protect_socket in create_udp function), leaf returned fake IP address instead of the domain real IP address so routing not works properly

what you think about it?, I think I need to make a local-dns-over-https server and send this kind of DNS requests to it, it's correct way!?

one question, by my understanding, leaf send requests asIs (like v2ray AsIs) mean leaf didn't resolve domains locally and send all domains directly to proxy servers and proxy servers will resolve those, is it correct?

**Comment by mokhtarabadi** [comment]
also, I think it,s true for other systems other than Android, for example in my Linux os when I set resolveDomain to true in routing I see these lines in the leaf log
```
[2021-06-30 16:06:11][DEBUG] looking up host www.example.com on 1.0.0.1:53
[2021-06-30 16:06:11][DEBUG] allocate 198.18.0.6 for www.example.com
[2021-06-30 16:06:11][DEBUG] return 1 ips (ttl 1) for www.example.com from 1.1.1.1:53 in 1ms
[2021-06-30 16:06:11][TRACE] ips for www.example.com:
CacheEntry {
    ips: [
        198.18.0.6,
    ],
    deadline: Instant {
        tv_sec: 1827,
        tv_nsec: 55339567,
    },
}
[2021-06-30 16:06:11][TRACE] re-matching with resolved ip [198.18.0.6] for [www.example.com]
[2021-06-30 16:06:11][DEBUG] [198.18.0.6] matches ip-cidr [0.0.0.0/0]
[2021-06-30 16:06:11][DEBUG] picked route [proxy_out] for 26.26.26.1:53602 -> www.example.com:443
```

`198.18.0.6` not the IP of `www.example.com`!

**Comment by eycorsican** [comment]
One thing to clarify is that the resolved IPs can only affect the routing process but not the destination of the request, as you can see we're making [a clone](https://github.com/eycorsican/leaf/blob/23dc682e34803f0c148a3f47e46c4bdb709827c4/leaf/src/app/router.rs#L461) of the session.

It's impossible to define what are the `real IP addresses` of a domain as you can get different IPs in different conditions, for example, sending DNS request from different  geolocations. I can understand what you mean by the `real IPs addresses` is something without modifying by a firewall, but you can never tell whether the response has been modified from our perspective.

DNS over HTTPS is one way to prevent the requests from being modified by intermediate boxes.



> If you don't want those dns requests to be sent out directly, just try to put a dispatcher into the dns client and retrieve a socket with that for sending queries.

What I've suggested above is to use DNS over whatever-proxy-protocol to send DNS requests through your proxy server, hopefully there isn't a malicious firewall there.

I personally would like to avoid this kind of hassles by using powerful domain rule collections, which can cover most of the poisoned domains.
